### ISSUE TITLE
How best to add units as in meters, seconds, etc. without confusion with racket/unit ?

### ISSUE THREAD

**Issue by AlexKnauth** [racket/typed-racket#180]
What would be the best way of adding support for units (as in meters, seconds, etc.) to typed racket, without conflicting with or confusing with units from `racket/unit`? How should the interface be designed to avoid that?

**Comment by dfeltey** [comment]
The conflict here is just the name right? I don't think there is other risk of confusion with `racket/unit`. Would something like `measures` or `dimension` or something like that make sense instead of `unit`?

**Comment by AlexKnauth** [comment]
Measure would make sense for the type of a number with a unit, but I was hoping to make the units be semi-first-class values at type check time. I don't like `dimension` for that though, because that would imply that it is about Length versus Time, instead of meters versus seconds. If I were to use dimensions, meters and kilometers would have the same dimension, but be different units. One thing I was thinking of was having `unit:` or `u`, or something similar, but that would cause more confusion, so I don't really know what to do.

**Comment by stamourv** [comment]
I agree that `dimension` is not 100% precise, but I think it would at least give the general idea (which the docs could make more precise). One could also imagine a "gradual" aspect to that axis of the type system, from TR, to TR with dimensions, to TR with units of measures. I suspect the middle point on that spectrum could be useful, and maybe lighterweight than the 3rd.

Other possibilities: `Measure-Unit`, go with `Measure` and call a value with a measure a `Quantity`, `Metric` (in the general sense of the term).

**Comment by samth** [comment]
I mostly agree with @stamourv. I think you should pick a term that isn't unit, such as dimension or measure, and just go with it.

**Comment by AlexKnauth** [comment]
I think I'll try it with measure-unit.

**Comment by jsmaniac** [comment]
`measure-unit` is a bit verbose, but at least it's clear.

I second what stamourv hinted, it would be great to have Meter a subtype of Length (but not of Time), and be able to write (* (Meter 200) (Kilometer 2)), and get as a result a Surface (equivalent to 400000 squared meter, but which needs conversion to be able to be used). Another function could take only Meter as a parameter, which would require (possibly explicit) conversion if we have a value in Kilometers. These sound hard to implement with the current type system, though.

**Comment by AlexKnauth** [comment]
I think that using `define-new-subtype` types (Distinction types), I could get somewhere, though what I have in mind would treat meters and kilometers as completely separate units that can't be added together. However, `(+ (Meters 200) (* (Kilometers 2) 1000m/1km))` would yield 2200 meters.
However I've been having some trouble; a program like
```racket
#lang typed/racket
(define-base-measure-unit m)
(define-type Real-Meters (Measure Real m))
```
Seems to hang or go into an infinite loop somewhere, and I have no idea how or why.

**Comment by jsmaniac** [comment]
Sorry, I deleted my comment after realizing that you had already implemented most of that anyway. I'm putting it back so the conversation makes more sense.

--

`measure-unit` is a bit verbose, but at least it makes things clear.

I second what stamourv said, it would be great to have `Meter` be a subtype of `Length` (but not of `Time`), and have `(* (Meters 200) (Kilometers 2))` give a `Surface` equivalent to 400000 squared meters, but which would need to be converted in order to be used. Another function could just take `Meter`, and passing it a value in `Kilometer`s would require implicit or explicit conversion. Implementing these sounds difficult with the current type system though.

--

To answer your comment about adding them, people could always define a `case→` function, like what is currently done for add (where `(+ Integer Integer)` gives Integer, but mixing types gives the more general Number):

```scheme
(: mixed-add (case→ (→ Meter Meter Meter)
                    (→ Kilometer Kilometer Kilometer)
                    (→ (U Meter Kilometer) (U Meter Kilometer) Length)))
```

**Comment by AlexKnauth** [comment]
Re: case-> and mixed-add:
That wouldn't work unless there was a runtime data structure holding the number and the unit.
But if that's what you're looking for, then that already exists:
http://pkg-build.racket-lang.org/doc/measures-with-dimensions/index.html
https://github.com/AlexKnauth/measures-with-dimensions
(see also https://github.com/Metaxal/measures)

**Comment by AlexKnauth** [comment]
Question:
Should I try extend the type-checking of +, -, *, and / expressions to support measure arithmetic, or would it be better to define my own versions that would support this.
Is it possible to express this behavior with a pure type, as opposed to another special case in the type-checker?

Here's what I have so far:
https://github.com/AlexKnauth/typed-racket/tree/dimensions

**Comment by samth** [comment]
I think adding new operations is better than adding more to `+`.

**Comment by AlexKnauth** [comment]
Ok. And do you know if it's possible to express measure arithmetic with types, or would there need to be special cases in the type-checker for this to work?

**Comment by stamourv** [comment]
On the one hand, extending `+` and co would make enriching code easier. On the other hand, these types are already ridiculously complicated, and I'd feel really bad for you if you had to go hack on them. :)

I think measure arithmetic should be expressible with types. If you look at what we've done with the numeric tower, it should be somewhat similar. This paper has more details: http://users.eecs.northwestern.edu/~stamourv/papers/numeric-tower.pdf

**Comment by samth** [comment]
There's a lot of literature on expressing units of measure in the type system. See papers, for example, by Anthony Kennedy about F# and by Eric Allen et al about Fortress.

**Comment by AlexKnauth** [comment]
Now I'm running into trouble with this bug in `mzlib/pconvert`:
```racket
#lang racket/base
(require mzlib/pconvert)
(print-convert (hash 1 2))
; '(make-hash (list (cons 1 2))), should be '(make-immutable-hash (list (cons 1 2)))
```

**Comment by samth** [comment]
Can you just report that there? It should be pretty easy to fix too.

**Comment by AlexKnauth** [comment]
Where is it? I see it's provided by `pconvert-lib`, but where is that?

**Comment by samth** [comment]
https://github.com/racket/pconvert-lib

**Comment by AlexKnauth** [comment]
Oh. I quickly browsed through the racket repos but I guess I missed it.
